src/server/worker/pipeline/transform/registry.go

```diff
@@ -0,0 +1,951 @@
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	pfsserver "github.com/pachyderm/pachyderm/src/server/pfs"
+	"github.com/pachyderm/pachyderm/src/server/pkg/backoff"
+	col "github.com/pachyderm/pachyderm/src/server/pkg/collection"
+	"github.com/pachyderm/pachyderm/src/server/pkg/errutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/obj"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsutil"
+	pfssync "github.com/pachyderm/pachyderm/src/server/pkg/sync"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
+	"github.com/pachyderm/pachyderm/src/server/pkg/work"
+	"github.com/pachyderm/pachyderm/src/server/worker/common"
+	"github.com/pachyderm/pachyderm/src/server/worker/datum"
+	"github.com/pachyderm/pachyderm/src/server/worker/driver"
+	"github.com/pachyderm/pachyderm/src/server/worker/logs"
+)
+
+const maxDatumsPerTask = 10000
+
+func jobHashtreesTag(jobID string) string {
+	return fmt.Sprintf("job-%s-hashtrees", jobID)
+}
+
+type pendingJob struct {
+	client        *client.APIClient
+	commitInfo    *pfs.CommitInfo
+	cancel        context.CancelFunc
+	logger        logs.TaggedLogger
+	index         int
+	orphan        bool
+	ji            *pps.JobInfo
+	jobData       *types.Any
+	datumsAdded   map[string]struct{}
+	datumsRemoved map[string]struct{}
+	datumTaskChan chan *work.Task
+	dit           datum.Iterator
+	hashtrees     []*HashtreeInfo
+}
+
+type registry struct {
+	driver      driver.Driver
+	logger      logs.TaggedLogger
+	workMaster  *work.Master
+	mutex       sync.Mutex
+	concurrency int
+	limiter     limit.ConcurrencyLimiter
+	datumsBase  map[string]struct{}
+	jobs        []*pendingJob
+}
+
+// Returns the registry or lazily instantiates it
+func newRegistry(
+	logger logs.TaggedLogger,
+	driver driver.Driver,
+) (*registry, error) {
+	// Determine the maximum number of concurrent tasks we will allow
+	concurrency, err := driver.GetExpectedNumWorkers()
+	if err != nil {
+		return nil, err
+	}
+
+	return &registry{
+		driver:      driver,
+		logger:      logger,
+		concurrency: concurrency,
+		workMaster:  driver.NewTaskMaster(),
+		limiter:     limit.New(concurrency),
+	}, nil
+}
+
+// finishJob will transactionally finish the output and meta commit for the job,
+// and optionally set the job's state.
+func (reg *registry) finishJob(pj *pendingJob) error {
+	// TODO: accept hashtrees for commits
+	_, err := pj.client.RunBatchInTransaction(func(builder *client.TransactionBuilder) error {
+		if pj.ji.StatsCommit != nil {
+			if _, err := builder.PfsAPIClient.FinishCommit(pj.client.Ctx(), &pfs.FinishCommitRequest{
+				Commit: pj.ji.StatsCommit,
+				Empty:  true,
+			}); err != nil {
+				return err
+			}
+		}
+
+		if _, err := builder.PfsAPIClient.FinishCommit(pj.client.Ctx(), &pfs.FinishCommitRequest{
+			Commit: pj.ji.OutputCommit,
+			Empty:  true,
+		}); err != nil {
+			return err
+		}
+
+		if _, err := builder.PpsAPIClient.UpdateJobState(pj.client.Ctx(), &pps.UpdateJobStateRequest{
```

Comment:
Correct me if I am wrong, but this is not implemented yet, right? It looks like batch transactions only apply to pfs operations, but I think the goal is to be able to batch this with pfs operations as well. This is exactly what we need to get ourselves out of the unfinished commit/job issues that have been plaguing us.